icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp

```diff
@@ -9,39 +16,404 @@ namespace icicle {
+        nof_hashes = nof_hashes * cur_layer.m_hash.input_default_chunk_size() / next_layer_total_size;
+      }
+    }
+
+    // build task and dispatch it to task manager
+    void dispatch_task(HashTask* task, int cur_layer_idx, const uint64_t cur_segment_idx, const std::byte* input_bytes)
+    {
+      // Calculate Next-Segment-ID. The current task generates inputs for Next-Segment
+      LayerDB& cur_layer = m_layers[cur_layer_idx];
+      const uint64_t next_layer_idx = cur_layer_idx + 1;
+
+      // Set HashTask input
+      const uint64_t input_offset =
+        (cur_layer_idx != 0) && (cur_layer_idx - 1 < m_output_store_min_layer)
+          ? 0
+          : cur_segment_idx * NOF_OPERATIONS_PER_TASK * cur_layer.m_hash.input_default_chunk_size();
+      task->m_input = &(input_bytes[input_offset]);
+
+      task->m_hash = cur_layer.m_hash;
+      task->m_hash_config = &cur_layer.m_last_hash_config;
+      task->m_layer_idx = cur_layer_idx;
+      task->m_segment_idx = cur_segment_idx;
+
+      // If this is the last layer
+      if (next_layer_idx == m_layers.size()) {
+        task->m_output = cur_layer.m_results.data();
+        task->dispatch();
+        return;
+      }
+
+      // This is not the last layer
+      const uint64_t next_input_size = m_layers[next_layer_idx].m_hash.input_default_chunk_size();
+      const uint64_t next_segment_idx = cur_segment_idx * cur_layer.m_hash.output_size() / next_input_size;
+      const uint64_t next_segment_id = next_segment_idx ^ (next_layer_idx << 56);
+
+      // If next_segment does not appear m_map_segment_id_2_inputs, add it
+      auto next_segment_it = m_map_segment_id_2_inputs.find(next_segment_id);
+      if (next_segment_it == m_map_segment_id_2_inputs.end()) {
+        const int size_to_allocate =
+          (cur_layer_idx < m_output_store_min_layer) ? NOF_OPERATIONS_PER_TASK * next_input_size : 0;
+        const auto result = m_map_segment_id_2_inputs.emplace(next_segment_id, new SegmentDB(size_to_allocate));
+        next_segment_it = result.first;
+      }
+
+      // calc task_output
+      const uint64_t task_output_offset = cur_segment_idx * NOF_OPERATIONS_PER_TASK * cur_layer.m_hash.output_size();
+      task->m_output =
+        (cur_layer_idx < m_output_store_min_layer)
+          ? &(next_segment_it->second->m_input_data[task_output_offset % (NOF_OPERATIONS_PER_TASK * next_input_size)])
+          : &(cur_layer.m_results[task_output_offset]);
+
+      // If this is not the last hash, update hash config
+      if ((cur_segment_idx + 1) * NOF_OPERATIONS_PER_TASK < cur_layer.m_nof_hashes)
+        task->m_hash_config = &cur_layer.m_hash_config;
+
+      // Set task next segment to handle return data
+      task->m_next_segment_idx = next_segment_idx;
+
+      // dispatch task
+      task->dispatch();
+    }
+
+    // restore the proof path from the tree
+    void copy_to_path_from_store_min_layer(const uint64_t input_chunk_offset, bool is_prooned, std::byte*& path) const
```

Comment:
the reference here is really confusing in terms of API. I would suggest returning the size as return value and let the caller increment the path pointer.

Reply:
done